File: RandomizedQuickSort.py
# Randomized Quick Sort using Hoare's partition Scheme

# worst case time complexity O(n*n)
# average case time complexity O(nlogn)
# better than quick sort with last element as pivot

import random


def quickSort(array, start, stop):
    if(start < stop):

        pivotindex = randPivot(array, start, stop)

        quickSort(array, start, pivotindex)
        quickSort(array, pivotindex + 1, stop)


# generate random pivot points
def randPivot(array, start, stop):

    # random number between stop/4 and 3*stop/4
    pivot = random.randrange(start, stop)

    array[start], array[pivot] = array[pivot], array[start]
    return partition(array, start, stop)


# make partitions based on the random pivot point
def partition(array, start, stop):
    pivot = start
    i = start - 1
    j = stop + 1
    while True:
        while True:
            i = i + 1
            if array[i] >= array[pivot]:
                break
        while True:
            j = j - 1
            if array[j] <= array[pivot]:
                break
        if i >= j:
            return j
        array[i], array[j] = array[j], array[i]


if __name__ == "__main__":
    arr = [10, 4, 63, 90, 43, 20, 1, 9]
    print(arr)
    quickSort(arr, 0, len(arr) - 1)
    print('array after randomized quick sort ', arr)

File: Sorting/randomizedQuickSort.cpp
// Randomized Quick Sort using Hoare's partition scheme

// worst case time complexity O(n*n)
// average time complexity O(nlogn)
// better than quick sort with last element as pivot

#include <iostream>
#include <stdlib.h>
#include <time.h>
#include <vector>
using namespace std;

// make partitions based on the random pivots
int partition(vector<int> &arr, int start, int stop) {
  int pivot = start, i = start - 1, j = stop + 1;

  while (true) {
    while (true) {
      i++;
      if (arr[i] >= arr[pivot])
        break;
    }
    while (true) {
      j--;
      if (arr[j] <= arr[pivot])
        break;
    }
    if (i >= j)
      return j;

    // swap elements at index i and j
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
  }
}

// generating random pivot points
int randPivot(vector<int> &arr, int start, int stop) {

  srand(time(NULL));
  int pivot = (rand() % (stop - start + 1)) + start;

  // swap numbers at index start and pivot
  int temp = arr[start];
  arr[start] = arr[pivot];
  arr[pivot] = temp;

  return partition(arr, start, stop);
}

// quick sort function
void quickSort(vector<int> &arr, int start, int stop) {
  if (start < stop) {
    int pivotindex = randPivot(arr, start, stop);

    quickSort(arr, start, pivotindex);
    quickSort(arr, pivotindex + 1, stop);
  }
}

// driver code
int main() {
  vector<int> array = {10, 4, 63, 90, 43, 20, 1, 9};

  cout << "unsorted array: ";
  for (auto x : array)
    cout << x << " ";
  cout << "\n";
  quickSort(array, 0, array.size() - 1);
  cout << "sorted array: ";
  for (auto x : array)
    cout << x << " ";
  cout << "\n";

  return 0;
}